man/conflict.man
.TH CONFLICT %manext8% "June 24, 2012" "%nmhversion%"
.\"
.\" %nmhwarning%
.\"
.SH NAME
conflict \- search for alias/password conflicts
.SH SYNOPSIS
.HP 5
.na
.B %libdir%/conflict
.RB [ \-search
.IR directory ]
.RB [ \-mail
.IR name ]
.RB [ \-version ]
.RB [ \-help ] 
.RI [ aliasfiles
\&...]
.ad
.SH DESCRIPTION
.B "Conflict is deprecated and will be removed from the next nmh release."
.PP
.B Conflict
is a program that checks to see if the interface between
.B nmh
and the transport system is in good shape
.PP
.B Conflict
also checks for maildrops in %mailspool% which do not
belong to a valid user.  It assumes that no user name will start with
`.', and thus ignores files in %mailspool% which begin with `.'.  It also
checks for entries in the
.IR group (5)
file which do not belong
to a valid user, and for users who do not have a valid group number.
In addition duplicate users and groups are noted.
.PP
If the
.B \-mail
.I name
switch is used, then the results will be sent
to the specified
.IR name .
Otherwise, the results are sent to the standard output.
.PP
The
.B \-search
.I directory
switch can be used to search directories
other than %mailspool% and to report anomalies in those directories.
The
.B \-search
.I directory
switch can appear more than one time in an
invocation to
.BR conflict .
.PP
.B Conflict
should be run under
.IR cron (8),
or whenever system accounting takes place.
.SH FILES
.fc ^ ~
.nf
.ta \w'%etcdir%/ExtraBigFileName  'u
^%etcdir%/mts.conf~^nmh mts configuration file
^/etc/passwd~^List of users
^/etc/group~^List of groups
^%bindir%/mhmail~^Program to send mail
^%mailspool%/~^Directory of mail drop
.fi
.SH "PROFILE COMPONENTS"
None
.SH "SEE ALSO"
.IR mh\-alias (5)
.SH "DEFAULTS"
.nf
.RB ` aliasfiles "' defaults to %etcdir%/MailAliases"
.fi
.SH CONTEXT
None


NEWS
#
# NEWS - Release notes for nmh 1.5
#

Welcome to nmh, the new version of the classic MH mail handling
system.  Long-time MH and nmh users should read carefully the NOTABLE
CHANGES section, as there are some subtle but significant changes
to the way nmh now behaves.  Otherwise, please see the README and INSTALL
files for help on getting started with nmh.

------------
NEW FEATURES
------------

The following are new features for the 1.5 release of nmh.

- Improved performance for TLS encryption of the SMTP protocol.
- MH-E and POP support are now always compiled in.
- The selection of a locking directory for dot-locking files is now
  controllable via configure.
- fileproc and mhlproc mh-profile are now obeyed by send, rcvdist,
  and whatnow.
- New mh-format instructions have been implemented: %(putlit),
  %(concataddr), %(myhost), %(myname), %(localmbox).  See mh-format(5)
  for more details.
- Nmh's idea of the local mailbox is now configurable via a new profile
  entry, Local-Mailbox.  See mh-profile(5) for more details.
- comp, forw, and dist now process all component templates through
  mh-format(5), and now all support new command line arguments -from,
  -to, -cc, -fcc, and -subject.
- burst(1) now supports any valid message delimiter according to RFC 934.
- post(8) now requires a From: header in all drafts that it processes;
  see NOTABLE CHANGES below for more details.
- post(8) can now explicitly control the SMTP envelope address either via
  the Sender: header or the new Envelope-From: header.
- The -attach option is now the default for send (using the header
  Nmh-Attachment) and -attachformat 1 is the default attachment type.
- repl and dist now support new switches -atfile and -noatfile to control
  the creation of the ./@ link to the message being replied or
  redistributed.  The current default is -atfile.
- Support (if available) for readline editing and filename completion at
  the WhatNow? prompt.
- Support for using an external command to filter the body of a message
  through in mhl.  See mhl(1) for more details, specifically, the
  "format" keyword, the "formatproc" entry in mh-profile(5), and the
  -fmtproc switch for mhl and repl.
- Preliminary support for improved MIME handling when replying to messages!
  Yes, a long requested feature has a solution.  A perl script
  called replyfilter is available; it is designed to act as a mhl
  external filter to process MIME messages in a more logical way.
  It is available in $(srcdir)/docs/contrib/replyfilter or is
  typically installed as $(prefix)/share/doc/nmh/contrib/replyfilter.
  See the comments at the top of replyfilter for usage information;
  it will likely require some adjustment for your site.  replyfilter
  requires the MIME-Tools and MailTools perl modules.

----------------------------
OBSOLETE/DEPRECATED FEATURES
----------------------------

The following features have either been removed completely or marked
as officially deprecated for removal in a future release of nmh.

- The creation of the ./@ link to the message being replied or
  redistributed is deprecated; -noatfile will be the default in the
  next release.  If there are no requests to maintain -atfile it
  will be removed in a future release.
- Old code enabled by the UCI preprocessor definition has been removed.
- All configuration options that were formerly only adjustable by editing
  config.h have either been moved into configure or removed completely.
- Support for the built-in ftp client (used when dealing with external-body
  message parts in MIME messages) has been removed.
- The following environment variables are marked as obsolete/deprecated.
  Support for them will be removed from the next nmh release:
    MHPOPDEBUG (use -snoop command line switch instead)
    MM_NOASK (use -nolist and -nopause command line switches instead)
    NOMHNPROC (use -nocheckmime command line switch instead)
    FACEPROC (undocumented faceproc feature to be removed)
- The hostable option of mts.conf is no longer needed and has been removed.

---------------
NOTABLE CHANGES
---------------

The largest change is that post(8) now requires a From: header in all drafts
that it processes.

The impetus for this change has been the evolution of email and client
systems.  When MH was first developed the norm was to have tightly managed
clients that were part of the same domain as a user's email address.  This
is no longer a valid assumption, and the algorithm used by nmh to determine
the user's local mailbox is no longer a reasonable default.

Unfortunately, the way that nmh uses the user's local mailbox used to happen
very late in the message process; it was handled by post during message
submission.  If no From: header exists in the current draft post would
insert one based on it's idea of the local mailbox, which was frequently
wrong.

After a long discussion within the nmh development community, it was
decided that the only reasonable way forward was to move the local mailbox
handling up a level and place it under more direct control of the end
user.  post(8) no longer will try to determine the local mailbox name;
it will take that information from the draft message.  Thus a From: header
is now required in all drafts.

As part of these changes, all message composition utilities now process
their component files using mh-format.  All of the default component
files have been modified to include a From: header in them; this will
give the user a chance to see nmh's idea of the local mailbox and modify
it if desired.  The new mh-profile(5) entry Local-Mailbox can be used
to adjust the local mailbox name if the system default is not appropriate.

So how should end-users adapt to these changes?  Here are some suggestions:

1) If you use the default component files, everything should basically work
   the same for you.  You should see a From: header in the drafts of your
   messages.  If the address you see in the From: header is incorrect

uip/conflict.c

/*
 * conflict.c -- check for conflicts in mail system
 *
 * This code is Copyright (c) 2002, by the authors of nmh.  See the
 * COPYRIGHT file in the root directory of the nmh distribution for
 * complete copyright information.
 */

#include <h/mh.h>
#include <fcntl.h>
#include <h/aliasbr.h>
#include <h/mts.h>
#include <h/utils.h>
#include <grp.h>
#include <pwd.h>

/*
 * maximum number of directories that can
 * be specified using -search switch.
 */
#define	NDIRS	100

/*
 * Add space for group names, 100 at a time
 */
#define	NGRPS	100

#define CONFLICT_SWITCHES \
    X("mail name", 0, MAILSW) \
    X("search directory", 0, SERCHSW) \
    X("version", 0, VERSIONSW) \
    X("help", 0, HELPSW) \

#define X(sw, minchars, id) id,
DEFINE_SWITCH_ENUM(CONFLICT);
#undef X

#define X(sw, minchars, id) { sw, minchars, id },
DEFINE_SWITCH_ARRAY(CONFLICT, switches);
#undef X

static char *mail = NULL;
static char *dirs[NDIRS];
static FILE *out = NULL;

extern struct aka  *akahead;
extern struct home *homehead;

/*
 * prototypes
 */
void alias_files (int, char **);
void pwd_names (void);
void grp_names (void);
void grp_members (void);
void grp_ids (void);
void maildrops (void);
void mdrop(char *);
int check (char *);
void setup (void);


int
main (int argc, char **argv)
{
    int	akp = 0, dp = 0;
    char *cp, **argp, **arguments;
    char buf[BUFSIZ], *akv[50];

#ifdef LOCALE
    setlocale(LC_ALL, "");
#endif
    invo_name = r1bindex (argv[0], '/');

    /* foil search of user profile/context */
    if (context_foil (NULL) == -1)
	done (1);

    mts_init (invo_name);
    arguments = getarguments (invo_name, argc, argv, 0);
    argp = arguments;

    while ((cp = *argp++)) {
	if (*cp == '-') {
	    switch (smatch (++cp, switches)) {
		case AMBIGSW: 
		    ambigsw (cp, switches);
		    done (1);
		case UNKWNSW: 
		    adios (NULL, "-%s unknown", cp);

		case HELPSW: 
		    snprintf (buf, sizeof(buf), "%s [switches] [aliasfiles ...]",
			invo_name);
		    print_help (buf, switches, 0);
		    done (0);
		case VERSIONSW:
		    print_version(invo_name);
		    done (0);

		case MAILSW: 
		    if (!(cp = *argp++) || *cp == '-')
			adios (NULL, "missing argument to %s", argp[-2]);
		    if (mail)
			adios (NULL, "mail to one address only");
		    else
			mail = cp;
		    continue;

		case SERCHSW: 
		    if (!(cp = *argp++) || *cp == '-')
			adios (NULL, "missing argument to %s", argp[-2]);
		    if (dp >= NDIRS)
			adios (NULL, "more than %d directories", NDIRS);
		    dirs[dp++] = cp;
		    continue;
	    }
	}
	akv[akp++] = cp;
    }

    if (akp == 0)
	akv[akp++] = AliasFile;
    if (!homehead)
	init_pw ();
    if (!mail)
	out = stdout;
    dirs[dp] = NULL;

    alias_files (akp, akv);
    pwd_names ();
    grp_names ();
    grp_members ();
    grp_ids ();
    maildrops ();

    done (0);
    return 1;
}


void
alias_files (int akp, char **akv)
{
    register int i, err;

    for (i = 0; i < akp; i++)
	if ((err = alias (akv[i])) != AK_OK) {
	    setup ();
	    fprintf (out, "aliasing error in %s - %s\n", akv[i], akerror (err));
	}
	else
	    if (out && !mail)
		fprintf (out, "alias file %s is ok\n", akv[i]);
}


void
pwd_names (void)
{
    int hit = 0;
    register struct home *hm, *lm;

    for (hm = homehead; hm; hm = hm->h_next)
	for (lm = hm->h_next; lm; lm = lm->h_next)
	    if (strcmp (hm->h_name, lm->h_name) == 0) {
		setup ();
		fprintf (out, "duplicate user %s(uid=%d)\n",
			lm->h_name, (int) lm->h_uid);
		hit++;
	    }

    if (!hit && out && !mail)
	fprintf (out, "no duplicate users\n");
}


void
grp_names (void)
{
    int numgroups, maxgroups;
    int i, hit = 0;
    char **grps;
    struct group *gr;

    /* allocate space NGRPS at a time */
    numgroups = 0;
    maxgroups = NGRPS;
    grps = (char **) mh_xmalloc((size_t) (maxgroups * sizeof(*grps)));

    setgrent ();
    while ((gr = getgrent ())) {
	for (i = 0; i < numgroups; i++)
	    if (!strcmp (grps[i], gr->gr_name)) {
		setup ();
		fprintf (out, "duplicate group %s(gid=%d)\n",
			gr->gr_name, (int) gr->gr_gid);
		hit++;
		break;
	    }
	if (i >= numgroups) {
	    if (numgroups >= maxgroups) {
		maxgroups += NGRPS;
		grps = (char **) mh_xrealloc(grps,
		    (size_t) (maxgroups * sizeof(*grps)));
	    }
	    grps[numgroups++] = getcpy (gr->gr_name);
	}
    }
    endgrent ();

    for (i = 0; i < numgroups; i++)
	free (grps[i]);
    free (grps);

    if (!hit && out && !mail)
	fprintf (out, "no duplicate groups\n");
}


void
grp_members (void)
{
    register int hit = 0;
    register char **cp, **dp;
    register struct group *gr;
    register struct home  *hm;

    setgrent ();
    while ((gr = getgrent ())) {
	for (cp = gr->gr_mem; *cp; cp++) {
	    for (hm = homehead; hm; hm = hm->h_next)
		if (!strcmp (*cp, hm->h_name))
		    break;
	    if (hm == NULL) {
		setup ();
		fprintf (out, "group %s(gid=%d) has unknown member %s\n",
			gr->gr_name, (int) gr->gr_gid, *cp);
		hit++;
	    } else {
		hm->h_ngrps++;
	    }

	    for (dp = cp + 1; *dp; dp++)
		if (strcmp (*cp, *dp) == 0) {
		    setup ();
		    fprintf (out, "group %s(gid=%d) has duplicate member %s\n",
			    gr->gr_name, (int) gr->gr_gid, *cp);
		    hit++;
		}
	}
    }
    endgrent ();

    for (hm = homehead; hm; hm = hm->h_next)
	if (hm->h_ngrps > NGROUPS_MAX) {
	    setup ();
	    fprintf (out, "user %s is a member of %d groups (max %d)\n",
		    hm->h_name, hm->h_ngrps, NGROUPS_MAX);
	    hit++;
	}

    if (!hit && out && !mail)
	fprintf (out, "all group members accounted for\n");
}


void
grp_ids (void)
{		/* -DRAND not implemented at most places */
    register int hit = 0;
    register struct home *hm;

    for (hm = homehead; hm; hm = hm->h_next)
	if (getgrgid (hm->h_gid) == NULL) {
	    setup ();
	    fprintf (out, "user %s(uid=%d) has unknown group-id %d\n",
		    hm->h_name, (int) hm->h_uid, (int) hm->h_gid);
	    hit++;
	}

    if (!hit && out && !mail)
	fprintf (out, "all group-id users accounted for\n");
}


void
maildrops (void) 
{
    register int i;

    if (mmdfldir && *mmdfldir)
	mdrop (mmdfldir);
    if (uucpldir && *uucpldir)
	mdrop (uucpldir);
    for (i = 0; dirs[i]; i++)
	mdrop (dirs[i]);
}


void
mdrop(char *drop)
{
    register int hit = 0;
    register struct dirent *dp;
    register DIR *dd = opendir (drop);

    if (!dd) {
	setup ();
	fprintf (out, "unable to open maildrop area %s\n", drop);
	return;
    }

    while ((dp = readdir (dd)))
	if (dp->d_name[0] != '.' && !check (dp->d_name)) {
	    setup ();
	    fprintf (out,
		    "there is a maildrop for the unknown user %s in %s\n",
		    dp->d_name, drop);
	    hit++;
	}

    closedir (dd);
    if (!hit && out && !mail)
	fprintf (out, "all maildrops accounted for in %s\n", drop);
}


int
check (char *s)
{
    register struct home *hm;

    for (hm = homehead; hm; hm = hm->h_next)
	if (!strcmp (s, hm->h_name))
	    return 1;
    return 0;
}

void
setup (void)
{
    int fd, pd[2];

    if (out)
	return;

    if (mail) {
	if (pipe (pd) == NOTOK)
	    adios ("pipe", "unable to");

	switch (fork ()) {
	    case NOTOK: 
		adios ("fork", "unable to");

	    case OK: 
		close (pd[1]);
		if (pd[0] != 0) {
		    dup2 (pd[0], 0);
		    close (pd[0]);
		}
		if ((fd = open ("/dev/null", O_WRONLY)) != NOTOK)
		    if (fd != 1) {
			dup2 (fd, 1);
			close (fd);
		    }
		execlp (mailproc, r1bindex (mailproc, '/'),
			mail, "-subject", invo_name, (void *) NULL);
		adios (mailproc, "unable to exec ");

	    default: 
		close (pd[0]);
		out = fdopen (pd[1], "w");
		fprintf (out, "%s: the following is suspicious\n\n",
			invo_name);
	}
    }
}
